### ISSUE TITLE
How to make predictions

### ISSUE THREAD

**Issue by opentld** [liznerski/fcdd#22]
> Another comment on this. If you want the raw pixel values of the heatmaps that are generated, you can find these in the "tims" directory and load them via `torch.load("heatmaps_paper_global_lbl0.pth")`. If you want a script that loads the snapshot of a model and generates pixel-wise anomaly score tensors, I added such an implementation [here](https://github.com/liznerski/fcdd/blob/master/python/fcdd/runners/run_prediction_with_snapshot.py). I hope this helps.

I have run run_prediction_with_snapshot.py,  and get the all_anomaly_scores tensor，but how to judge whether it is abnormal through pixel-wise anomaly score tensors?  and how to get the heat map through the run_prediction_with_snapshot.py method?

@liznerski

_Originally posted by @opentld in https://github.com/liznerski/fcdd/issues/15#issuecomment-955158932_

another question：
I only put one picture,and the path is shown in the figure:
![image](https://user-images.githubusercontent.com/21096515/139523266-cfe52439-6029-4f73-878a-c0a822bbb843.png)
 then I ran python runners/run_custom.py --supervise-mode unsupervised --it 1 -e 0 --load ../../data/test/model/snapshot.pt --net FCDD_CNN224_VGG_F
Get the following result：
![image](https://user-images.githubusercontent.com/21096515/139523344-e2bc3460-f39f-4526-96f0-c7ecaad2ebe5.png)
![image](https://user-images.githubusercontent.com/21096515/139523358-043ac6e8-09b3-41c1-b68b-6ebbf869b109.png)
heatmaps_global.png as shown:
![heatmaps_global](https://user-images.githubusercontent.com/21096515/139523378-f837021b-aa86-4bfc-bf3c-b63a75d9c68e.png)
heatmaps_paper_global_lbl0.png:
![heatmaps_paper_global_lbl0](https://user-images.githubusercontent.com/21096515/139523421-3ac1f18d-60e6-42c1-b2de-fd38c9052693.png)
heatmaps_paper_global_lbl1.png:
![heatmaps_paper_global_lbl1](https://user-images.githubusercontent.com/21096515/139523427-a69069ad-c1b5-4c73-9cff-f8ffbc8e9e15.png)
heatmaps_paper_local_lbl0.png:
![heatmaps_paper_local_lbl0](https://user-images.githubusercontent.com/21096515/139523434-17cbb0d0-b3d5-4af1-ab1d-7e0ec25c236e.png)
heatmaps_paper_local_lbl1.png:
![heatmaps_paper_local_lbl1](https://user-images.githubusercontent.com/21096515/139523442-563454a3-ce9f-4e86-97c1-f42ef7835b45.png)
train_heatmaps_global.png:
![train_heatmaps_global](https://user-images.githubusercontent.com/21096515/139523448-202cb5d4-eaa5-4195-8620-b7aeb1812dfc.png)

so, why does run_custom.py try to process all the classes? For example, I did not place the cable picture, but it still generates normal_1? I just want to perform anomaly detection on the bottle image I set, and get its classification result and heat map. How can I do it?

_Originally posted by @opentld in https://github.com/liznerski/fcdd/issues/15#issuecomment-955161257_

@liznerski

**Comment by liznerski** [comment]
> I have run run_prediction_with_snapshot.py, and get the all_anomaly_scores tensor，but how to judge whether it is abnormal through pixel-wise anomaly score tensors? and how to get the heat map through the run_prediction_with_snapshot.py method?

To transform the pixel-wise anomaly scores to a sample-wise anomaly score, use: 
```
trainer.reduce_ascore(all_anomaly_scores)
```

The heatmaps require a bit more work. You need to define a logger, the variance for receptive upsampling, and a quantile for normalization:
```
logger = Logger("../../data/results/foo/")
quantile = 0.97 
gauss_std = 8 
trainer = FCDDTrainer(net, None, None, (None, None), logger, 'fcdd', gauss_std, quantile, 224)
[...]
trainer.heatmap_generation(all_labels, all_anomaly_scores, all_inputs)
```
However, note that the "global normalization" heatmaps might be different because they are normalized w.r.t. all samples passed to the method, which in your case might be just a few instead of the whole dataset for the respective class. Also, you might need to tune the quantile a bit for best visualization (looking at your provided heatmaps).  


> I only put one picture,and the path is shown in the figure: [...].
then I ran python runners/run_custom.py --supervise-mode unsupervised --it 1 -e 0 --load ../../data/test/model/snapshot.pt --net FCDD_CNN224_VGG_F
Get the following result：[...].

I assume you put a dummy image in the respective train folder? Otherwise, you should have had an error that it could not find valid files in the train path. 

Even with this dummy image, you should have had an error because you miss another directory specifying the label per class (since you didn't set the [ovr](https://github.com/liznerski/fcdd/blob/master/python/fcdd/runners/run_custom.py#L16) tag). That is, instead of `datasets/custom/test/bottle/004.png`, you need `datasets/custom/test/bottle/anomalous/004.png`. The error should have been raised [here](https://github.com/liznerski/fcdd/blob/master/python/fcdd/datasets/image_folder.py#L192) since [this](https://github.com/liznerski/fcdd/blob/master/python/fcdd/datasets/image_folder.py#L205) line should have mapped all labels to the anomalous label. I don't know why that didn't happen for you.

Further, even with the correct directory structure, testing should fail because it cannot compute an AUC with just one label. You need to, again, put a dummy image for the other label; i.e., `datasets/custom/test/bottle/normal/00x.png`. However, if the image is not a valid normal instance, the AUC will obviously be nonsense. 

Unfortunately, I am not able to reproduce your results. I guess that, e.g., you have specified the wrong [datapath](https://github.com/liznerski/fcdd/blob/master/python/fcdd/runners/argparse_configs.py#L30)? If you only have one image in your dataset folder, the trainer should have no way to access this number of images. 

In general, I recommend using the [mvtec runner](https://github.com/liznerski/fcdd/blob/master/python/fcdd/runners/run_mvtec.py) for your purposes. You can change [this](https://github.com/liznerski/fcdd/blob/master/python/fcdd/training/bases.py#L242) line to get more heatmaps (e.g., 30): 
```
self.heatmap_generation(
    labels, anomaly_scores, imgs, gtmaps, grads,
    specific_idx=specific_viz_ids, show_per_cls=30
)
```
And if you want specific test heatmaps, you can use the specific_idx parameter (i.e., the [viz-ids](https://github.com/liznerski/fcdd/blob/master/python/fcdd/runners/argparse_configs.py#L34) argument for the runner).

**Comment by opentld** [comment]
I modified the run_prediction_with_snapshot.py file as follows:

# Create a trainer to use its loss function for computing anomaly scores
ds = ImageFolder(images_path, transform, transforms.Lambda(lambda x: 0))
loader = DataLoader(ds, batch_size=16, num_workers=0)

logger = Logger("../../data/results/foo/")
quantile = 0.97 
gauss_std = 8

#trainer = FCDDTrainer(net, None, None, (None, None), None, 'fcdd', 8, 0.99, 128)  # these parameters will have no effect (used for heatmaps)
trainer = FCDDTrainer(net, None, None, (None, None), None, 'fcdd', gauss_std, quantile, 224)  # these parameters will have no effect (used for heatmaps)

trainer.load(snapshot)

all_anomaly_scores = []
all_inputs = []
all_labels = []

for inputs, labels in loader:
    all_inputs.append(inputs)
    all_labels.append(labels)

    inputs = inputs.cuda()
    with torch.no_grad():
        outputs = trainer.net(inputs)
        anomaly_scores = trainer.anomaly_score(trainer.loss(outputs, inputs, labels, reduce='none'))
        anomaly_scores = trainer.net.receptive_upsample(anomaly_scores, reception=True, std=8, cpu=False)
        all_anomaly_scores.append(anomaly_scores.cpu())
        
# all_anomaly_scores will be a tensor containing pixel-wise anomaly scores for all images
all_anomaly_scores = torch.cat(all_anomaly_scores)
all_inputs = torch.cat(all_inputs)
all_labels = torch.cat(all_labels).int()

trainer.heatmap_generation(all_labels, all_anomaly_scores, all_inputs)

scores = trainer.reduce_ascore(all_anomaly_scores)
print(scores)

But the following error occurred：
![image](https://user-images.githubusercontent.com/21096515/140293703-c071167c-102d-4d46-9660-cff9386c4201.png)

If I don’t try to generate a heat map, but just calculate trainer.reduce_ascore(all_anomaly_scores), I get scores like this
![image](https://user-images.githubusercontent.com/21096515/140293827-8a8a8c01-de7a-47e1-98b6-81246d875827.png)
How should I understand these scores? It seems difficult to use a simple threshold segmentation to determine whether it is an anomaly?

@liznerski

**Comment by opentld** [comment]
In other words, I want to load a trained model, such as bottle’s snapshot.pt, and then use it to predict a picture, use its score to determine whether it is abnormal, and generate a heat map of this picture. So what should I do?

@liznerski

**Comment by liznerski** [comment]
You forgot to add `logger` to the trainer's parameters. 
```
trainer = FCDDTrainer(net, None, None, (None, None), logger, 'fcdd', gauss_std, quantile, 224) 
```
Also, the `heatmap_generation` method expects the labels to be a list and not a tensor: 
```
trainer.heatmap_generation(all_labels.tolist(), all_anomaly_scores, all_inputs)
```
But that's both not what caused your error. Again, I cannot reproduce this. The line your error is referring to ([478](https://github.com/liznerski/fcdd/blob/windows/python/fcdd/training/bases.py#L478)) is part of the documentation in current code, so nothing that could raise an error. Did you make any changes or use old code? 

Anyway, I've implemented some changes to make the `run_prediction_with_snapshot.py` script also create heatmaps. See 9ded37dad4f5d4bec29738f646b1d960662ee037. This script should exactly serve your purposes. I've tested this, and it should work (at least on Linux). 

> How should I understand these scores? It seems difficult to use a simple threshold segmentation to determine whether it is an anomaly?

Well, the larger the score, the more the model deems the sample anomalous. To find a threshold you would need to set the score into context by comparing it to scores for known nominal samples (training samples). 

Hope this helps ;)

**Comment by opentld** [comment]
I use the Windows branch. After updating the code, it can run normally.
The problem now is that I put 209 positive samples under the train/good folder in the dataset bottle and a negative sample e000.png，into the foo folder, 
![image](https://user-images.githubusercontent.com/21096515/140482962-470cea02-28fe-4cb7-8cc5-aec7d51e3b14.png)
and try to run run_prediction_with_snapshot.py to make predictions. The results are as follows:
![image](https://user-images.githubusercontent.com/21096515/140483053-1ca06f98-753e-4354-9ba2-80aea5451643.png)
How can I determine the threshold to make a judgment? It seems that the score of this negative sample is not high.. .Moreover, observing the heat map, it seems that this negative sample is not different from the positive sample.
Am I doing it correctly? Or what should I do?

In addition, what should I do if I want to display the heat map of a sample separately?

@liznerski

**Comment by liznerski** [comment]
Oh. Yes, thanks for your report. Didn't notice. We forgot to set the network to eval mode. I pushed a fix (7af3d8eadabee81ab8f7db5dea7f8389ef090213). Should work now. For my snapshot, the anomaly scores are around 0.008 for nominal samples and 0.19 for anomalous ones.

**Comment by opentld** [comment]
After adding eval, the result is correct. Thank you for your great works！
There is another question. What should I do if I only display the heat map of the abnormal sample? For example, I only test an abnormal sample and display its heat map?

@liznerski

**Comment by liznerski** [comment]
The `run_prediction_with_snapshot.py` script should be fine with having just one image in the respective folder. Though, with only one image, there will be a "local" heatmap only (the "global" one will actually be "local" too since there are no other images to compare with).

**Comment by opentld** [comment]
![image](https://user-images.githubusercontent.com/21096515/140709416-429a9ce9-de19-4f1d-9847-2a6ebdb1c67a.png)

I tested all positive samples, the maximum value is 0.0495, which corresponds to this picture：

![141](https://user-images.githubusercontent.com/21096515/140709470-7a3b2ff3-fb34-466b-a5a9-8c2e913f5900.png)

It seems that the ring shadow caused the higher score，you can also think of it as an anomaly ;)

**Comment by opentld** [comment]
https://github.com/liznerski/fcdd/issues/2#issuecomment-963045086

**Comment by liznerski** [comment]
Commented in #2
